crates/hc_sandbox/src/zome_call.rs

```diff
@@ -0,0 +1,516 @@
+            .filter_map(|app_interface| {
+                if app_interface.installed_app_id.is_some()
+                    && app_interface.installed_app_id.as_ref().unwrap() != &installed_app_id
+                {
+                    return None;
+                }
+
+                if app_interface.allowed_origins.is_allowed("sandbox") {
+                    Some(app_interface.port)
+                } else {
+                    None
+                }
+            })
+            .next();
+
+        let port = match existing_port {
+            Some(port) => port,
+            None => {
+                let response = cmd_runner
+                    .command(AdminRequest::AttachAppInterface {
+                        port: None,
+                        allowed_origins: AllowedOrigins::Origins(
+                            vec!["sandbox".to_string()].into_iter().collect(),
+                        ),
+                        installed_app_id: None,
+                    })
+                    .await?;
+
+                match response {
+                    AdminResponse::AppInterfaceAttached { port } => port,
+                    other => anyhow::bail!(
+                        "Unexpected response while attaching app interface: {:?}",
+                        other
+                    ),
+                }
+            }
+        };
+
+        let (ws_send, ws_recv) = connect(
+            Arc::new(WebsocketConfig::CLIENT_DEFAULT),
+            ConnectRequest::new(
+                format!("localhost:{port}")
+                    .to_socket_addrs()
+                    .unwrap()
+                    .next()
+                    .unwrap(),
+            )
+            .try_set_header("origin", "sandbox")
+            .unwrap(),
+        )
+        .await?;
+
+        let _recv = WsPollRecv::new::<AppResponse>(ws_recv);
+
+        let response = cmd_runner
+            .command(AdminRequest::IssueAppAuthenticationToken(
+                IssueAppAuthenticationTokenPayload::for_installed_app_id(installed_app_id.clone()),
+            ))
+            .await
+            .context("Could not issue app authentication token")?;
+
+        let token = match response {
+            AdminResponse::AppAuthenticationTokenIssued(issued) => issued.token,
+            other => anyhow::bail!(
+                "Unexpected response while issuing app authentication token: {:?}",
+                other
+            ),
+        };
+
+        ws_send
+            .authenticate(AppAuthenticationRequest { token })
+            .await
+            .context("Failed to authenticate app interface connection")?;
+
+        Ok(Self { ws_send, _recv })
+    }
+
+    async fn request(&self, request: AppRequest) -> anyhow::Result<AppResponse> {
+        Ok(self.ws_send.request(request).await?)
+    }
+}
+
+/// You do not need to do anything with this type. While it is held it will keep polling a websocket
+/// receiver.
+struct WsPollRecv(tokio::task::JoinHandle<()>);
```

Comment:
Similarly, this exists in several places. I wonder if we should provide it in the `holochain_websocket` crate?